---
title: "ChIP-seq analyses"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(fastqcr)
library(dplyr)
```

Variables for shell scripts are stored in the file "shell_variables_ChIP-seq.sh" and can be modified as needed:

```{bash variables}
cat shell_variables_ChIP-seq.sh
```

#QC of sequence reads

Technical quality of library construction and sequencing run was tested using fastQC

```{bash}
fastqc --version
```

```{bash, eval=FALSE}
source shell_variables_ChIP-seq.sh

mkdir -p "${fastQC_dir}"
ls -1 "${fastq_dir}" | grep "${fastq_file_ext}" > "${fastq_file_list}"
fastq_files=`cat "${fastq_file_list}"`
for i in ${fastq_files};
do
  fastqc --quiet --outdir "${fastQC_dir}" --threads "${CPU}" "${fastq_dir}${i}"
done
```

Summarize fastQC results

```{r, message=FALSE, eval=FALSE}
fastQC_dir <- './fastQC_ChIP-seq/'
qc <- qc_aggregate(fastQC_dir, progressbar = FALSE)
qc_report(qc.path = fastQC_dir, result.file = paste0(fastQC_dir, 'fastQC-summary'),
          experiment = 'RNAP ChIP-seq in Bacillus subtilis (torpedo)')
```

Results of QC:

* 50 nt reads

* 45-48M reads per input sample, 31-40M reads per IP sample

* no adapter contamination

* good read quality scores

* cleaning/trimming of reads is not needed

#Read mapping

Genome sequence and annotation:

* Bacillus subtilis subps. subtilis strain 168 chromosome, complete genome (gi|255767013|ref|NC_000964|); https://www.ncbi.nlm.nih.gov/nuccore/NC_000964

* genome annotation (GFF): genome-build-accession NCBI_Assembly:GCF_000009045.1; ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF_000009045.1_ASM904v1 downloaded 15/Nov/2015

```{bash}
hisat2 --version
samtools --version
```

```{bash, eval=FALSE}
source shell_variables_ChIP-seq.sh

# build HISAT2 index
hisat2-build --quiet -p "${CPU}" "${genome}" "${genome}"

# map
mkdir -p "${bam_dir}"
fastq_files=`cat "${fastq_file_list}"`
for i in ${fastq_files};
do
	infile="${fastq_dir}${i}"
	outfile="${bam_dir}${i}.bam"
	echo "${outfile}"
	hisat2 -x "${genome}" -U "${infile}" --threads "${CPU}" --no-spliced-alignment --quiet \
	  --met-file "${outfile}.log" | samtools view -b -q "${Q}" --threads "${CPU}" - | samtools sort -o "${outfile}" - 
	samtools index "${outfile}"
	samtools view -c --threads "${CPU}" "${outfile}"
done
```

Notes

* multimapping and poorly mapped reads were removed from further analysis (MAPQ <10)

#Create genome coverage files

```{bash}
deeptools --version
```

```{bash}
source shell_variables_ChIP-seq.sh

mkdir -p "${coverage_dir}"
ls -1 "${bam_dir}" | grep ".bam$" > "${bam_file_list}"
mapfile -t bam_files < "${bam_file_list}"
sample_count=$(expr "${#bam_files[@]}" / 2)

# get genome coverage for all input files -> inspect in IGV for irregularities
for i in $(eval echo "{1..$sample_count}");
do
	input_file="${bam_dir}${bam_files[${i}-1]}"
	bamCoverage -p "${CPU}" -b "${input_file}" -o "${coverage_dir}${bam_files[${i}-1]}.bw"
done

# normalize IP samples to the corresponding input samples (read count-normalized ratio)
for i in $(eval echo "{1..$sample_count}");
do
	input_file="${bam_dir}${bam_files[${i}-1]}"
	ip_file="${bam_dir}${bam_files[${i}+${sample_count}-1]}"
	bamCompare --operation ratio --binSize 1 -p "${CPU}" --scaleFactorsMethod readCount -b1 "${ip_file}" -b2 "${input_file}" -o "${coverage_dir}${bam_files[${i}+${sample_count}-1]}${normalized_suffix}.bw"
done

# average replicate data
normalized_ko_bw_files=`ls -1 -d "${coverage_dir}"/* | grep "${normalized_suffix}" | grep "${ko}"`
normalized_wt_bw_files=`ls -1 -d "${coverage_dir}"/* | grep "${normalized_suffix}" | grep "${wt}"`
multiBigwigSummary bins --outRawCounts "${coverage_dir}${ko}${normalized_suffix}.tab" --binSize 1 -p "${CPU}" -o "${coverage_dir}${ko}${normalized_suffix}.npz" -b ${normalized_ko_bw_files}
multiBigwigSummary bins --outRawCounts "${coverage_dir}${wt}${normalized_suffix}.tab" --binSize 1 -p "${CPU}" -o "${coverage_dir}${wt}${normalized_suffix}.npz" -b ${normalized_wt_bw_files}
```

Notes on input samples

* 16S and 23S rRNA genes showed increased coverage

* no coverage on rna"X" and rrn"X" genes

```{r}
coverage_dir <- './coverage/'
ko <- "KO"
wt <- "wt"

coverage_ko <- read.delim(file = list.files(coverage_dir, full.names = TRUE,
                                            pattern = paste0('.*', ko, '.*tab$')),
                          stringsAsFactors = FALSE)
coverage_ko <- cbind(coverage_ko[, 1:3], rowMeans(coverage_ko[, 4:ncol(coverage_ko)], na.rm = TRUE))
colnames(coverage_ko) <- c('chr', 'start', 'end', 'coverage')
coverage_ko <- coverage_ko[order(coverage_ko$start), ]
writeLines('track type=bedGraph name=KO', 
           con = paste0(coverage_dir, 'KO.input-normalized.mean.bedgraph'))
write.table(coverage_ko, file = paste0(coverage_dir, 'KO.input-normalized.mean.bedgraph'), 
            quote = FALSE, append = TRUE, sep = '\t', row.names = FALSE, col.names = FALSE)

coverage_wt <- read.delim(file = list.files(coverage_dir, full.names = TRUE,
                                            pattern = paste0('.*', wt, '.*tab$')),
                          stringsAsFactors = FALSE)
coverage_wt <- cbind(coverage_wt[, 1:3], rowMeans(coverage_wt[, 4:ncol(coverage_wt)], na.rm = TRUE))
colnames(coverage_wt) <- colnames(coverage_ko)
coverage_wt <- coverage_wt[order(coverage_wt$start), ]
writeLines('track type=bedGraph name=wt', 
           con = paste0(coverage_dir, 'wt.input-normalized.mean.bedgraph'))
write.table(coverage_wt, file = paste0(coverage_dir, 'wt.input-normalized.mean.bedgraph'), 
            quote = FALSE, append = TRUE, sep = '\t', row.names = FALSE, col.names = FALSE)
```

***

```{r}
sessionInfo()
```
