---
title: "ChIP-seq analyses"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(fastqcr)
library(dplyr)
```

Variables for shell scripts are stored in the file "shell_variables_ChIP-seq.sh" and can be modified as needed:

```{bash variables}
cat shell_variables_ChIP-seq.sh
```

#Genome sequence and annotation

* Bacillus subtilis subps. subtilis strain 168 chromosome, complete genome (gi|255767013|ref|NC_000964|); https://www.ncbi.nlm.nih.gov/nuccore/NC_000964

* genome annotation (GFF): genome-build-accession NCBI_Assembly:GCF_000009045.1; ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF_000009045.1_ASM904v1 (downloaded 15/Nov/2015)

The original GFF file has been modified:

1) the chromosome name was changed from NC_000964.3 to gi|255767013|ref|NC_000964.3| in order to match the chromosome name in the genome FASTA file

2) the strand orientation was swapped for all features to get correct read counts from stranded RNA-seq libraries

3) the `summarizeOverlaps()` function works with gene names (attribute Name), which for some reason did not work with this GFF file (zero counts were reported for about 1/2 of all genes). However, the unique systematic names (attribute locus_tag) work correctly. Therefore the values of Name and locus_tag were swapped

#QC of sequence reads

Technical quality of library construction and sequencing run was tested using fastQC

```{bash}
fastqc --version
```

```{bash, eval=FALSE}
source shell_variables_ChIP-seq.sh

mkdir -p "${fastQC_dir}"
ls -1 "${fastq_dir}" | grep "${fastq_file_ext}" > "${fastq_file_list}"
fastq_files=`cat "${fastq_file_list}"`
for i in ${fastq_files};
do
  fastqc --quiet --outdir "${fastQC_dir}" --threads "${CPU}" "${fastq_dir}${i}"
done
```

Summarize fastQC results

```{r, message=FALSE, eval=FALSE}
fastQC_dir <- './fastQC_ChIP-seq/'
qc <- qc_aggregate(fastQC_dir, progressbar = FALSE)
qc_report(qc.path = fastQC_dir, result.file = paste0(fastQC_dir, 'fastQC-summary'),
          experiment = 'RNAP ChIP-seq in Bacillus subtilis (torpedo)')
```

Results of QC:

* 50 nt reads

* 45-48M reads per input sample, 31-40M reads per IP sample

* no adapter contamination

* good read quality scores

* cleaning/trimming of reads is not needed

#Read mapping into B. subtilis genome

```{bash}
hisat2 --version
samtools --version
```

```{bash, eval=FALSE}
source shell_variables_ChIP-seq.sh

# build HISAT2 index
hisat2-build --quiet -p "${CPU}" "${genome}" "${genome}"

# map
mkdir -p "${bam_dir}"
fastq_files=`cat "${fastq_file_list}"`
for i in ${fastq_files};
do
	infile="${fastq_dir}${i}"
	outfile="${bam_dir}${i}.bam"
	echo "${outfile}"
	hisat2 -x "${genome}" -U "${infile}" --threads "${CPU}" --no-spliced-alignment --quiet --met-file "${outfile}.log" | samtools view -b -q "${Q}" --threads "${CPU}" - | samtools sort -o "${outfile}" - 
	samtools index "${outfile}"
	samtools view -c --threads "${CPU}" "${outfile}"
done
```

Notes

* multimapping and poorly mapped reads were removed from further analysis (MAPQ <10)

#Create genome coverage files

```{bash}
deeptools --version
```

```{bash}
source shell_variables_ChIP-seq.sh

mkdir -p "${coverage_dir}"
ls -1 "${bam_dir}" | grep ".bam$" > "${bam_file_list}"
mapfile -t bam_files < "${bam_file_list}"
sample_count=$(expr "${#bam_files[@]}" / 2)

# get genome coverage for all input files -> inspect in IGV for irregularities
for i in $(eval echo "{1..$sample_count}");
do
	input_file="${bam_dir}${bam_files[${i}-1]}"
	bamCoverage -p "${CPU}" -b "${input_file}" -o "${coverage_dir}${bam_files[${i}-1]}.bw"
done

# normalize IP samples to the corresponding input samples (read count-normalized ratio)
for i in $(eval echo "{1..$sample_count}");
do
	input_file="${bam_dir}${bam_files[${i}-1]}"
	ip_file="${bam_dir}${bam_files[${i}+${sample_count}-1]}"
	bamCompare --operation ratio --binSize 1 -p "${CPU}" --scaleFactorsMethod readCount -b1 "${ip_file}" -b2 "${input_file}" -o "${coverage_dir}${bam_files[${i}+${sample_count}-1]}${normalized_suffix}.bw"
done

# average replicate data
normalized_ko_bw_files=`ls -1 -d "${coverage_dir}"/* | grep "${normalized_suffix}" | grep "${ko}"`
normalized_wt_bw_files=`ls -1 -d "${coverage_dir}"/* | grep "${normalized_suffix}" | grep "${wt}"`
multiBigwigSummary bins --outRawCounts "${coverage_dir}${ko}${normalized_suffix}.tab" --binSize 1 -p "${CPU}" -o "${coverage_dir}${ko}${normalized_suffix}.npz" -b ${normalized_ko_bw_files}
multiBigwigSummary bins --outRawCounts "${coverage_dir}${wt}${normalized_suffix}.tab" --binSize 1 -p "${CPU}" -o "${coverage_dir}${wt}${normalized_suffix}.npz" -b ${normalized_wt_bw_files}
```

Notes on input samples

* 16S and 23S rRNA genes showed increased coverage

* no coverage on rna"X" and rrn"X" genes

```{r}
coverage_dir <- './coverage_ChIP-seq/'
ko <- "KO"
wt <- "wt"

coverage_ko <- read.delim(file = list.files(coverage_dir, full.names = TRUE,
                                            pattern = paste0('.*', ko, '.*tab$')),
                          stringsAsFactors = FALSE)
coverage_ko <- cbind(coverage_ko[, 1:3], rowMeans(coverage_ko[, 4:ncol(coverage_ko)], na.rm = TRUE))
colnames(coverage_ko) <- c('chr', 'start', 'end', 'coverage')
coverage_ko <- coverage_ko[order(coverage_ko$start), ]
writeLines('track type=bedGraph name=KO', 
           con = paste0(coverage_dir, 'KO.input-normalized.mean.bedgraph'))
write.table(coverage_ko, file = paste0(coverage_dir, 'KO.input-normalized.mean.bedgraph'), 
            quote = FALSE, append = TRUE, sep = '\t', row.names = FALSE, col.names = FALSE)

coverage_wt <- read.delim(file = list.files(coverage_dir, full.names = TRUE,
                                            pattern = paste0('.*', wt, '.*tab$')),
                          stringsAsFactors = FALSE)
coverage_wt <- cbind(coverage_wt[, 1:3], rowMeans(coverage_wt[, 4:ncol(coverage_wt)], na.rm = TRUE))
colnames(coverage_wt) <- colnames(coverage_ko)
coverage_wt <- coverage_wt[order(coverage_wt$start), ]
writeLines('track type=bedGraph name=wt', 
           con = paste0(coverage_dir, 'wt.input-normalized.mean.bedgraph'))
write.table(coverage_wt, file = paste0(coverage_dir, 'wt.input-normalized.mean.bedgraph'), 
            quote = FALSE, append = TRUE, sep = '\t', row.names = FALSE, col.names = FALSE)
```

#Create coverage plots for selected genes

Plots produced manually from IGV 2.6.3

```{bash}
source shell_variables_ChIP-seq.sh

mkdir -p "${image_dir}"
```

# Calculate per-gene coverage

```{r, include=FALSE}
# functions for GFF manipulation were obtained from:
# https://stat.ethz.ch/pipermail/bioconductor/2008-October/024669.html
getAttributeField <- function (x, field, attrsep = ";") {
  s = strsplit(x, split = attrsep, fixed = TRUE)
  sapply(s, function(atts) {
    a = strsplit(atts, split = "=", fixed = TRUE)
    m = match(field, sapply(a, "[", 1))
    if (!is.na(m)) {
      rv = a[[m]][2]
    }
    else {
      rv = as.character(NA)
    }
    return(rv)
  })
}

gffRead <- function(gffFile, nrows = -1) {
  cat("Reading ", gffFile, ": ", sep="")
  gff = read.table(gffFile, sep="\t", as.is=TRUE, quote="",
                   header=FALSE, comment.char="#", nrows = nrows,
                   colClasses=c("character", "character", "character", "integer",  
                                "integer",
                                "character", "character", "character", "character"))
  colnames(gff) = c("seqname", "source", "feature", "start", "end",
                    "score", "strand", "frame", "attributes")
  cat("found", nrow(gff), "rows with classes:",
      paste(sapply(gff, class), collapse=", "), "\n")
  stopifnot(!any(is.na(gff$start)), !any(is.na(gff$end)))
  return(gff)
}
```

```{r}
gff_file <- './genome/GCF_000009045.1_ASM904v1_genomic_reverse_SYS.gff'
gff <- gffRead(gff_file)
gff <- gff[gff$feature == 'gene', ]
gff$locus_tag <- getAttributeField(gff$attributes, field = 'locus_tag') # trivial gene name
gff$Name <- getAttributeField(gff$attributes, field = 'Name') # systematic gene ID (BSUxxxxx)

gene_coverage <- data.frame(seqname = gff$seqname, 
                            start = as.numeric(gff$start),
                            end = as.numeric(gff$end), 
                            locus_tag = gff$locus_tag, 
                            Name = gff$Name, 
                            cov_wt = as.numeric(NA), 
                            cov_ko = as.numeric(NA),
                            stringsAsFactors = FALSE)
for (i in 1:nrow(gene_coverage)){
  gene_length_kb <- (gene_coverage[i, 'end'] - gene_coverage[i, 'start'] + 1) / 1000
  gene_coverage[i, 'cov_wt'] <- 
    sum(coverage_wt[which(coverage_wt$chr == gene_coverage[i, 'seqname'] &
                          coverage_wt$start + 1 >=  gene_coverage[i, 'start'] &
                          coverage_wt$end <= gene_coverage[i, 'end']), 'coverage']) / gene_length_kb
  
  gene_coverage[i, 'cov_ko'] <- 
    sum(coverage_ko[which(coverage_ko$chr == gene_coverage[i, 'seqname'] &
                          coverage_ko$start + 1 >=  gene_coverage[i, 'start'] &
                          coverage_ko$end <= gene_coverage[i, 'end']), 'coverage']) / gene_length_kb
#  print(nrow(gene_coverage) - i)
}
write.table(gene_coverage, file = paste0(coverage_dir, 'ChIP-seq_gene_coverage.txt'),
            quote = FALSE, sep = '\t', row.names = FALSE, col.names = TRUE)
```

#Comparisons of ChIP-seq and RNA-seq data

```{r}
#library(RColorBrewer)
gene_coverage <- gene_coverage[gene_coverage$Name %in% rownames(counts.len.norm), ]
gene_coverage <- gene_coverage[order(gene_coverage$Name), ]

load('/home/martin/Drive/lab/experiments/PÅ™FUK_postdoc/18 - RNA Pol mutants in B. subtilis/count_tables.rda')
pseudocount <- 0
pdf('./images/Figure2.pdf')

par(mar = c(5, 5, 5, 5), lwd = 2, cex = 1.5)
options(scipen = -1)
#limits <- range(counts.len.norm) + pseudocount
limits <- range(gene_coverage[, c('cov_wt', 'cov_ko')]) + pseudocount
rbPal <- colorRampPalette(c('yellow', 'red'))

plot(gene_coverage$cov_ko, counts.len.norm[, 'ko.rna'] + pseudocount,
     main = '',
     xlab = '',
     ylab = '',
     log = 'xy', xlim = c(10, limits[2]), ylim = c(0.01, limits[2]),
     type = 'n')
abline(v = c(1, 10, 100, 1000, 10000), h = c(0.01, 1, 100, 10000), col = 'gray')
par(new = TRUE)
plot(gene_coverage$cov_ko, counts.len.norm[, 'ko.rna'] + pseudocount,
     main = 'rnjA KO\npooled 3 replicates\n(coloured by expression in WT)',
     xlab = 'RNAP occupancy at genes [norm. cov.]',
     ylab = 'gene RNA abundance [RPKM]',
     log = 'xy', xlim = c(10, limits[2]), ylim = c(0.01, limits[2]), pch = 20,
     col = rbPal(100)[as.numeric(cut(log2(counts.len.norm[, 'wt.rna'] + 0.01), breaks = 100))])

plot(gene_coverage$cov_wt, counts.len.norm[, 'wt.rna'] + pseudocount,
     main = '',
     xlab = '',
     ylab = '',
     log = 'xy', xlim = c(10, limits[2]), ylim = c(0.01, limits[2]),
     type = 'n')
abline(v = c(1, 10, 100, 1000, 10000), h = c(0.01, 1, 100, 10000), col = 'gray')
par(new = TRUE)
plot(gene_coverage$cov_wt, counts.len.norm[, 'wt.rna'] + pseudocount,
     main = 'WT\npooled 3 replicates\n(coloured by expression in WT)',
     xlab = 'RNAP occupancy at genes [norm. cov.]',
     ylab = 'gene RNA abundance [RPKM]',
     log = 'xy', xlim = c(10, limits[2]), ylim = c(0.01, limits[2]), pch = 20,
     col = rbPal(100)[as.numeric(cut(log2(counts.len.norm[, 'wt.rna'] + 0.01), breaks = 100))])
dev.off()
```

***

```{r}
sessionInfo()
```