---
title: "RNA-seq analyses"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(fastqcr)
library(dplyr)
library(GenomicAlignments)
library(GenomicFeatures)
library(DESeq2)
library(RColorBrewer)
library(gplots)
```

Variables for shell scripts are stored in the file "shell_variables_RNA-seq.sh" and can be modified as needed:

```{bash variables}
cat shell_variables_RNA-seq.sh
```

#Genome sequence and annotation

* Bacillus subtilis subps. subtilis strain 168 chromosome, complete genome (gi|255767013|ref|NC_000964|); https://www.ncbi.nlm.nih.gov/nuccore/NC_000964

* genome annotation (GFF): genome-build-accession NCBI_Assembly:GCF_000009045.1; ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF_000009045.1_ASM904v1 (downloaded 15/Nov/2015)

The original GFF file has been modified:

1) the chromosome name was changed from NC_000964.3 to gi|255767013|ref|NC_000964.3| in order to match the chromosome name in the genome FASTA file

2) the strand orientation was swapped for all features to get correct read counts from stranded RNA-seq libraries

3) the `summarizeOverlaps()` function works with gene names (attribute Name), which for some reason did not work with this GFF file (zero counts were reported for about 1/2 of all genes). However, the unique systematic names (attribute locus_tag) work correctly. Therefore the values of Name and locus_tag were swapped

#QC of sequence reads

Technical quality of library construction and sequencing run was tested using fastQC

```{bash}
fastqc --version
```

```{bash, eval=FALSE}
source shell_variables_RNA-seq.sh

mkdir -p "${fastQC_dir}"
ls -1 "${fastq_dir}" | grep "${fastq_file_ext}" > "${fastq_file_list}"
fastq_files=`cat "${fastq_file_list}"`
for i in ${fastq_files};
do
  fastqc --quiet --outdir "${fastQC_dir}" --threads "${CPU}" "${fastq_dir}${i}"
done
```

Summarize fastQC results

```{r, message=FALSE, eval=FALSE}
fastQC_dir <- './fastQC_RNA-seq/'
qc <- qc_aggregate(fastQC_dir, progressbar = FALSE)
qc_report(qc.path = fastQC_dir, result.file = paste0(fastQC_dir, 'fastQC-summary'),
          experiment = 'rnjA wt/KO RNA-seq in Bacillus subtilis (torpedo)')
```

Results of QC:

* 51 nt reads

* 14-28M reads per sample

* some samples contain overrepresented sequences but no adapter contamination detected

* good read quality scores

* cleaning/trimming of reads is not needed

#Read mapping into B. subtilis genome

```{bash}
hisat2 --version
samtools --version
```

```{bash, eval=FALSE}
source shell_variables_RNA-seq.sh

# build HISAT2 index
hisat2-build --quiet -p "${CPU}" "${genome}" "${genome}"

# map
mkdir -p "${bam_dir}"
fastq_files=`cat "${fastq_file_list}"`
for i in ${fastq_files};
do
	infile="${fastq_dir}${i}"
	outfile="${bam_dir}${i}.bam"
	echo "${outfile}"
	hisat2 -x "${genome}" -U "${infile}" --threads "${CPU}" --no-spliced-alignment --rna-strandness R --quiet --met-file "${outfile}.log" | samtools view -b -q "${Q}" --threads "${CPU}" - | samtools sort -o "${outfile}" - 
	samtools index "${outfile}"
	samtools view -c --threads "${CPU}" "${outfile}"
done
```

Notes

* multimapping and poorly mapped reads were removed from further analysis (MAPQ <10)

#Create genome coverage files

```{bash}
deeptools --version
```

```{bash, eval=FALSE}
source shell_variables_RNA-seq.sh

mkdir -p "${coverage_dir}"
ls -1 "${bam_dir}" | grep ".bam$" > "${bam_file_list}"
mapfile -t bam_files < "${bam_file_list}"
sample_count="${#bam_files[@]}"

# get genome coverage for all BAM files -> inspect in IGV for irregularities
for i in $(eval echo "{1..$sample_count}");
do
	input_file="${bam_dir}${bam_files[${i}-1]}"
	bamCoverage --normalizeUsing CPM --binSize 1 -p "${CPU}" -b "${input_file}" -o "${coverage_dir}${bam_files[${i}-1]}.bw"
done

# average replicate data
ko_bw_files=`ls -1 -d "${coverage_dir}"/* | grep "${ko}"`
wt_bw_files=`ls -1 -d "${coverage_dir}"/* | grep "${wt}"`
multiBigwigSummary bins --outRawCounts "${coverage_dir}${ko}.tab" --binSize 1 -p "${CPU}" -o "${coverage_dir}${ko}.npz" -b ${ko_bw_files}
multiBigwigSummary bins --outRawCounts "${coverage_dir}${wt}.tab" --binSize 1 -p "${CPU}" -o "${coverage_dir}${wt}.npz" -b ${wt_bw_files}
```

```{r, eval=TRUE}
coverage_dir <- './coverage_RNA-seq/'
ko <- "KO"
wt <- "WT"

coverage_ko <- read.delim(file = list.files(coverage_dir, full.names = TRUE,
                                            pattern = paste0('.*', ko, '.*tab$')),
                          stringsAsFactors = FALSE)
coverage_ko <- cbind(coverage_ko[, 1:3], rowMeans(coverage_ko[, 4:ncol(coverage_ko)], na.rm = TRUE))
colnames(coverage_ko) <- c('chr', 'start', 'end', 'coverage')
coverage_ko <- coverage_ko[order(coverage_ko$start), ]
writeLines('track type=bedGraph name=KO', 
           con = paste0(coverage_dir, 'KO.mean.bedgraph'))
write.table(coverage_ko, file = paste0(coverage_dir, 'KO.mean.bedgraph'), 
            quote = FALSE, append = TRUE, sep = '\t', row.names = FALSE, col.names = FALSE)

coverage_wt <- read.delim(file = list.files(coverage_dir, full.names = TRUE,
                                            pattern = paste0('.*', wt, '.*tab$')),
                          stringsAsFactors = FALSE)
coverage_wt <- cbind(coverage_wt[, 1:3], rowMeans(coverage_wt[, 4:ncol(coverage_wt)], na.rm = TRUE))
colnames(coverage_wt) <- colnames(coverage_ko)
coverage_wt <- coverage_wt[order(coverage_wt$start), ]
writeLines('track type=bedGraph name=wt', 
           con = paste0(coverage_dir, 'wt.mean.bedgraph'))
write.table(coverage_wt, file = paste0(coverage_dir, 'wt.mean.bedgraph'), 
            quote = FALSE, append = TRUE, sep = '\t', row.names = FALSE, col.names = FALSE)
```

# Calculate per-gene coverage

```{r, include=FALSE}
# functions for GFF manipulation were obtained from:
# https://stat.ethz.ch/pipermail/bioconductor/2008-October/024669.html
getAttributeField <- function (x, field, attrsep = ";") {
  s = strsplit(x, split = attrsep, fixed = TRUE)
  sapply(s, function(atts) {
    a = strsplit(atts, split = "=", fixed = TRUE)
    m = match(field, sapply(a, "[", 1))
    if (!is.na(m)) {
      rv = a[[m]][2]
    }
    else {
      rv = as.character(NA)
    }
    return(rv)
  })
}

gffRead <- function(gffFile, nrows = -1) {
  cat("Reading ", gffFile, ": ", sep="")
  gff = read.table(gffFile, sep="\t", as.is=TRUE, quote="",
                   header=FALSE, comment.char="#", nrows = nrows,
                   colClasses=c("character", "character", "character", "integer",  
                                "integer",
                                "character", "character", "character", "character"))
  colnames(gff) = c("seqname", "source", "feature", "start", "end",
                    "score", "strand", "frame", "attributes")
  cat("found", nrow(gff), "rows with classes:",
      paste(sapply(gff, class), collapse=", "), "\n")
  stopifnot(!any(is.na(gff$start)), !any(is.na(gff$end)))
  return(gff)
}
```

```{r, eval=TRUE}
gff_file <- './genome/GCF_000009045.1_ASM904v1_genomic_reverse_SYS.gff'
gff <- gffRead(gff_file)
gff <- gff[gff$feature == 'gene', ]
gff$locus_tag <- getAttributeField(gff$attributes, field = 'locus_tag') # trivial gene name
gff$Name <- getAttributeField(gff$attributes, field = 'Name') # systematic gene ID (BSUxxxxx)

gene_coverage <- data.frame(seqname = gff$seqname, 
                            start = as.numeric(gff$start),
                            end = as.numeric(gff$end), 
                            locus_tag = gff$locus_tag, 
                            Name = gff$Name, 
                            cov_wt = as.numeric(NA), 
                            cov_ko = as.numeric(NA),
                            stringsAsFactors = FALSE)
for (i in 1:nrow(gene_coverage)){
  gene_length_kb <- (gene_coverage[i, 'end'] - gene_coverage[i, 'start'] + 1) / 1000
  gene_coverage[i, 'cov_wt'] <- 
    sum(coverage_wt[which(coverage_wt$chr == gene_coverage[i, 'seqname'] &
                          coverage_wt$start + 1 >=  gene_coverage[i, 'start'] &
                          coverage_wt$end <= gene_coverage[i, 'end']), 'coverage']) / gene_length_kb
  
  gene_coverage[i, 'cov_ko'] <- 
    sum(coverage_ko[which(coverage_ko$chr == gene_coverage[i, 'seqname'] &
                          coverage_ko$start + 1 >=  gene_coverage[i, 'start'] &
                          coverage_ko$end <= gene_coverage[i, 'end']), 'coverage']) / gene_length_kb
#  print(nrow(gene_coverage) - i)
}
write.table(gene_coverage, file = paste0(coverage_dir, 'RNA-seq_gene_coverage.txt'),
            quote = FALSE, sep = '\t', row.names = FALSE, col.names = TRUE)
```

#Analysis of differential gene expression

```{r, eval=TRUE}
# construct count table
bam_dir <- './BAM_RNA-seq/'
bam <- BamFileList(list.files(bam_dir, pattern = "\\.bam$", full.names = TRUE),
                   asMates = FALSE,
                   yieldSize = 100000)
annot <- makeTxDbFromGFF(file = gff_file,
                         format = "gff3",
                         dataSource = 'NCBI_Assembly:GCF_000009045.1',
                         organism = 'Bacillus subtilis subsp. subtilis str. 168',
                         circ_seqs ="gi|255767013|ref|NC_000964.3|")
se <- summarizeOverlaps(genes(annot),
                      bam,
                      mode = 'Union',
                      singleEnd = TRUE,
                      ignore.strand = FALSE)

# prepare summarized experiment metadata
sampleInfo <- data.frame(cbind(NA, 
                               list.files(bam_dir, full.names = TRUE, pattern = '.bam$'),
                               c('KO', 'KO', 'KO', 'WT', 'WT', 'WT'),
                               c(1, 2, 3, 1, 2, 3)))
colnames(sampleInfo) <- c('sample', 'bam.file', 'genotype', 'run')
sampleInfo$sample <- paste(sampleInfo$genotype, sampleInfo$run, sep = '_')
colnames(se) <- sampleInfo$sample
colData(se) <- cbind(colData(se), sampleInfo)

# diagnose the dataset and identify differentially expressed genes (DEGs)
padj.threshold <- 0.05
dds <- DESeqDataSet(se, design = ~ run + genotype)
dds$genotype <- relevel(dds$genotype, 'WT')
dds <- DESeq(dds)

rld <- rlog(dds)
sampleDist <- dist(t(assay(rld)))
sampleDistMatrix <- as.matrix(sampleDist)
colnames(sampleDistMatrix) <- NULL
colours <- colorRampPalette(rev(brewer.pal(9, 'Blues')))(255)
heatmap.2(sampleDistMatrix, trace = 'none', col = colours, mar = c(9, 9))
print(plotPCA(rld, intgroup = 'genotype'))

res <- results(dds, alpha = padj.threshold)
summary(res)
for (i in rownames(res)){
  res[i, 7] <- gff[which(gff$Name == i), 'locus_tag']
}
colnames(res)[7] <- 'gene_name'
write.csv(res, file = 'DESeq2results_rnjAKO_vs_WT.txt', quote = FALSE)
res.sig <- res[which(res$padj <= padj.threshold), ]
res.sig <- res.sig[order(res.sig$log2FoldChange), ]
write.csv(res.sig, file = 'DESeq2results_rnjAKO_vs_WT.SIG.txt', quote = FALSE)
```

#Comparison of expression values determined from normalized coverage vs. by DESeq2

```{r, fig.asp=1}
gene_coverage <- read.delim('coverage_RNA-seq/RNA-seq_gene_coverage.txt', 
                            header = TRUE, stringsAsFactors = FALSE)
gene_coverage <- cbind(gene_coverage, log2(gene_coverage$cov_ko / gene_coverage$cov_wt))
colnames(gene_coverage)[ncol(gene_coverage)] <- 'log2foldChange'
gene_coverage <- gene_coverage[!is.na(gene_coverage$log2foldChange) &
                               !is.infinite(gene_coverage$log2foldChange) &
                               !is.nan(gene_coverage$log2foldChange), ]
res <- read.csv('DESeq2results_rnjAKO_vs_WT.txt', 
                header = TRUE, stringsAsFactors = FALSE, row.names = 1)
res <- res[complete.cases(res$log2FoldChange), ]
res <- res[which(rownames(res) %in% gene_coverage$Name), ]
res <- res[order(rownames(res)), ]
gene_coverage <- gene_coverage[which(gene_coverage$Name %in% rownames(res)), ]
gene_coverage <- gene_coverage[order(gene_coverage$Name), ]
plot(log2(gene_coverage$cov_ko / gene_coverage$cov_wt), res$log2FoldChange,
     pch = 20, col = '#00000050',
     xlab = 'log2(normalized coverage) [deeptools]',
     ylab = 'log2(fold change expression) [DESeq2]')
abline(a = -1, b = 1, col = 'grey')
abline(a = 1, b = 1, col = 'grey')
text(x = -5, y = 5,
     labels = paste0('Pearson\'s R = ', round(cor(x = log2(gene_coverage$cov_ko / 
                                                           gene_coverage$cov_wt),
                                               y = res$log2FoldChange,                         
                                               method = 'pearson'), 3)))
```

***

```{r}
sessionInfo()
```